File: src-tauri/src/tunnel.rs
use crate::config::AppConfig;
use std::fs::File;
use std::io::{Read, Write};
use std::net::TcpStream;
use std::path::PathBuf;
use std::process::{Child, Command, Stdio};
use std::sync::{Arc, Mutex};
use std::time::Duration;
use tauri::{command, AppHandle, Emitter, Manager, State};

pub struct TunnelState {
    pub child: Mutex<Option<Child>>,
    pub monitor_running: Mutex<bool>,
    pub config: Mutex<Option<AppConfig>>,
}

fn ssh_log_path(app: &AppHandle) -> PathBuf {
    app.path()
        .app_log_dir()
        .unwrap_or_else(|_| PathBuf::from("/tmp"))
        .join("csm-ssh.log")
}

fn kill_child_group(child: &mut Child) {
    let pid = child.id() as i32;
    let _ = child.kill();
    let _ = child.wait();
    // Ensure the entire process group is terminated (SSH may fork children)
    unsafe {
        let _ = libc::kill(-pid, libc::SIGKILL);
    }
}

pub fn kill_tunnel(state: &TunnelState) {
    let mut guard = state.child.lock().unwrap();
    if let Some(mut child) = guard.take() {
        kill_child_group(&mut child);
    }
}

fn kill_process_on_port(port: u16) {
    let output = std::process::Command::new("lsof")
        .args(["-ti", &format!(":{}", port)])
        .output();
    if let Ok(out) = output {
        let pids = String::from_utf8_lossy(&out.stdout);
        for pid in pids.lines().filter(|s| !s.is_empty()) {
            let _ = std::process::Command::new("kill")
                .args(["-9", pid])
                .status();
        }
    }
}

pub fn start_tunnel_inner(
    config: &AppConfig,
    state: &TunnelState,
    app: &AppHandle,
) -> Result<(), String> {
    // 1. If the port is healthy, reuse the existing tunnel
    if http_health_check(config.local_port) {
        println!("[TAURI] Port {} is healthy, reusing existing tunnel.", config.local_port);
        return Ok(());
    }

    // 2. Port is not healthy — kill any stale process and rebuild tunnel
    println!("[TAURI] Port {} is not healthy, rebuilding tunnel...", config.local_port);

    // 2. Stop any tunnel tracked by our own state
    {
        let mut guard = state.child.lock().map_err(|e| e.to_string())?;
        if let Some(mut child) = guard.take() {
            kill_child_group(&mut child);
        }
    }

    // 3. Aggressively clear any stale process holding the local port
    kill_process_on_port(config.local_port);
    std::thread::sleep(Duration::from_millis(300));

    let log_path = ssh_log_path(app);
    if let Some(parent) = log_path.parent() {
        let _ = std::fs::create_dir_all(parent);
    }
    let log_file = File::create(&log_path)
        .map_err(|e| format!("Failed to create ssh log file: {}", e))?;

    let mut cmd = Command::new("ssh");
    cmd.arg("-N")
        .arg("-o")
        .arg("ServerAliveInterval=30")
        .arg("-o")
        .arg("ServerAliveCountMax=3")
        .arg("-o")
        .arg("TCPKeepAlive=yes")
        .arg("-o")
        .arg("ExitOnForwardFailure=yes")
        .arg("-o")
        .arg("BatchMode=yes")
        .arg("-o")
        .arg("StrictHostKeyChecking=accept-new")
        .arg("-o")
        .arg("ConnectTimeout=15")
        .arg("-L")
        .arg(format!(
            "{}:localhost:{}",
            config.local_port, config.remote_csm_port
        ))
        .arg("-p")
        .arg(config.ssh_port.to_string());

    if let Some(identity) = &config.identity_file {
        if !identity.is_empty() {
            cmd.arg("-i").arg(identity);
        }
    }

    cmd.arg(format!("{}@{}", config.ssh_user, config.ssh_host));
    cmd.stdout(Stdio::null()).stderr(Stdio::from(log_file));

    // Run SSH in its own process group so we can kill the whole tree
    unsafe {
        use std::os::unix::process::CommandExt;
        cmd.pre_exec(|| {
            libc::setpgid(0, 0);
            Ok(())
        });
    }

    let child = cmd
        .spawn()
        .map_err(|e| format!("Failed to start ssh tunnel: {}", e))?;

    let mut guard = state.child.lock().map_err(|e| e.to_string())?;
    *guard = Some(child);

    // Wait briefly then verify the process did not exit immediately
    std::thread::sleep(Duration::from_secs(1));
    if let Some(ref mut c) = *guard {
        match c.try_wait() {
            Ok(Some(status)) => {
                let log_content = std::fs::read_to_string(&log_path).unwrap_or_default();
                let hint = if log_content.contains("Address already in use") {
                    "Local port is still occupied after cleanup. Try changing the Local Port in Settings (e.g., 18081)."
                } else if log_content.contains("Connection refused") {
                    "Remote SSH server refused the connection. Check SSH Host and Port."
                } else if log_content.contains("Permission denied") || log_content.contains("authentication") {
                    "SSH authentication failed. Check your SSH key (Identity File) or user name."
                } else {
                    ""
                };
                return Err(format!(
                    "SSH tunnel exited immediately (status: {}).\nLog:\n{}\n{}",
                    status, log_content, hint
                ));
            }
            Ok(None) => {}
            Err(e) => return Err(format!("Failed to check tunnel status: {}", e)),
        }
    }

    Ok(())
}

#[command]
pub fn start_tunnel(
    config: AppConfig,
    state: State<TunnelState>,
    app: AppHandle,
) -> Result<(), String> {
    start_tunnel_inner(&config, &state, &app)?;
    // Save config for monitor
    {
        let mut guard = state.config.lock().map_err(|e| e.to_string())?;
        *guard = Some(config);
    }
    Ok(())
}

#[command]
pub fn stop_tunnel(state: State<TunnelState>) -> Result<(), String> {
    stop_monitor(&state);
    kill_tunnel(&state);
    {
        let mut guard = state.config.lock().map_err(|e| e.to_string())?;
        *guard = None;
    }
    Ok(())
}

#[command]
pub fn start_tunnel_monitor(
    state: State<'_, TunnelState>,
    app: AppHandle,
) -> Result<(), String> {
    let config: AppConfig;
    {
        let guard = state.config.lock().map_err(|e| e.to_string())?;
        config = guard.clone().ok_or("No tunnel config found. Connect first.")?;
    }

    // Stop any existing monitor
    stop_monitor(&state);

    // Start new monitor
    {
        let mut guard = state.monitor_running.lock().map_err(|e| e.to_string())?;
        *guard = true;
    }

    std::thread::spawn(move || {
        monitor_tunnel(config, app);
    });

    println!("[TAURI] Tunnel monitor started");
    Ok(())
}

fn http_health_check(local_port: u16) -> bool {
    let addr = format!("127.0.0.1:{}", local_port);
    match TcpStream::connect_timeout(
        &addr.parse().expect("valid socket addr"),
        Duration::from_secs(2),
    ) {
        Ok(mut stream) => {
            let _ = stream.set_write_timeout(Some(Duration::from_secs(2)));
            let _ = stream.set_read_timeout(Some(Duration::from_secs(2)));
            let req = format!("HEAD /api/sessions HTTP/1.1\r\nHost: localhost:{}\r\nConnection: close\r\n\r\n", local_port);
            if stream.write_all(req.as_bytes()).is_err() {
                return false;
            }
            let mut buf = [0u8; 32];
            match stream.read(&mut buf) {
                Ok(n) if n > 0 => {
                    let response = String::from_utf8_lossy(&buf[..n]);
                    response.contains("HTTP/1.1 200") || response.contains("HTTP/1.1 401")
                }
                _ => false,
            }
        }
        Err(_) => false,
    }
}

#[command]
pub fn check_connection(local_port: u16) -> bool {
    http_health_check(local_port)
}

pub fn monitor_tunnel(
    config: AppConfig,
    app: AppHandle,
) {
    let mut consecutive_failures = 0u32;
    let mut rebuild_delay = 2u64;
    loop {
        std::thread::sleep(Duration::from_secs(10));

        if http_health_check(config.local_port) {
            if consecutive_failures > 0 {
                println!("[TAURI] Tunnel health check recovered");
                let _ = app.emit("tunnel-reconnected", ());
            }
            consecutive_failures = 0;
            rebuild_delay = 2;
            continue;
        }

        consecutive_failures += 1;
        println!("[TAURI] Tunnel health check failed ({}/2)", consecutive_failures);

        if consecutive_failures >= 2 {
            println!("[TAURI] Tunnel deemed dead, auto-rebuilding...");
            let _ = app.emit("tunnel-disconnected", ());
            consecutive_failures = 0;

            let state = app.state::<TunnelState>();
            std::thread::sleep(Duration::from_secs(rebuild_delay));
            match start_tunnel_inner(&config, &state, &app) {
                Ok(()) => {
                    println!("[TAURI] Tunnel rebuilt successfully");
                    let _ = app.emit("tunnel-reconnected", ());
                    rebuild_delay = 2;
                }
                Err(e) => {
                    println!("[TAURI] Tunnel rebuild failed: {}", e);
                    rebuild_delay = std::cmp::min(rebuild_delay * 2, 60);
                }
            }
        }
    }
}

pub fn stop_monitor(state: &TunnelState) {
    let mut guard = state.monitor_running.lock().unwrap();
    *guard = false;
}

File: scripts/tunnel.js
#!/usr/bin/env node
/**
 * CSM Tunnel Manager - Lightweight SSH tunnel daemon for Claude Session Manager
 *
 * Usage:
 *   node scripts/tunnel.js start    # foreground mode (logs to console)
 *   node scripts/tunnel.js daemon   # background mode (logs to ~/.csm/tunnel.log)
 *   node scripts/tunnel.js stop     # stop running tunnel
 *   node scripts/tunnel.js status   # show tunnel status
 *   node scripts/tunnel.js config   # interactive config setup
 */

const { spawn, exec } = require('child_process');
const fs = require('fs');
const path = require('path');
const os = require('os');
const http = require('http');

const CSM_DIR = path.join(os.homedir(), '.csm');
const CONFIG_PATH = path.join(CSM_DIR, 'tunnel.json');
const LOG_PATH = path.join(CSM_DIR, 'tunnel.log');
const PID_PATH = path.join(CSM_DIR, 'tunnel.pid');

function log(msg) {
  const line = `[${new Date().toISOString()}] ${msg}`;
  console.log(line);
  try {
    fs.appendFileSync(LOG_PATH, line + '\n');
  } catch {}
}

function loadConfig() {
  try {
    return JSON.parse(fs.readFileSync(CONFIG_PATH, 'utf8'));
  } catch {
    return null;
  }
}

function saveConfig(config) {
  if (!fs.existsSync(CSM_DIR)) fs.mkdirSync(CSM_DIR, { recursive: true });
  fs.writeFileSync(CONFIG_PATH, JSON.stringify(config, null, 2));
}

function expandHome(p) {
  if (p && p.startsWith('~/')) return path.join(os.homedir(), p.slice(2));
  return p;
}

function ask(question) {
  return new Promise((resolve) => {
    process.stdout.write(question + ' ');
    process.stdin.once('data', (data) => resolve(data.toString().trim()));
  });
}

async function setupConfig() {
  const config = loadConfig() || {};
  const host = await ask(`SSH Host${config.ssh_host ? ` [${config.ssh_host}]` : ''}:`);
  if (host) config.ssh_host = host;
  const user = await ask(`SSH User${config.ssh_user ? ` [${config.ssh_user}]` : ''}:`);
  if (user) config.ssh_user = user;
  const port = await ask(`SSH Port [${config.ssh_port || 22}]:`);
  if (port) config.ssh_port = parseInt(port, 10);
  const remote = await ask(`Remote CSM Port [${config.remote_csm_port || 8080}]:`);
  if (remote) config.remote_csm_port = parseInt(remote, 10);
  const local = await ask(`Local Port [${config.local_port || 18080}]:`);
  if (local) config.local_port = parseInt(local, 10);
  const identity = await ask(`Identity File [${config.identity_file || '~/.ssh/id_rsa'}]:`);
  if (identity) config.identity_file = identity;

  saveConfig(config);
  log('Config saved to ' + CONFIG_PATH);
  process.exit(0);
}

function buildSshArgs(config) {
  const args = [
    '-N',
    '-o', 'ServerAliveInterval=30',
    '-o', 'ServerAliveCountMax=3',
    '-o', 'TCPKeepAlive=yes',
    '-o', 'ExitOnForwardFailure=yes',
    '-o', 'BatchMode=yes',
    '-o', 'StrictHostKeyChecking=accept-new',
    '-o', 'ConnectTimeout=15',
    '-L', `${config.local_port}:localhost:${config.remote_csm_port}`,
    '-p', String(config.ssh_port || 22),
  ];
  if (config.identity_file) {
    args.push('-i', expandHome(config.identity_file));
  }
  args.push(`${config.ssh_user}@${config.ssh_host}`);
  return args;
}

function httpHealthCheck(port) {
  return new Promise((resolve) => {
    const req = http.request({
      host: '127.0.0.1',
      port,
      path: '/api/sessions',
      method: 'HEAD',
      timeout: 3000,
    }, (res) => {
      resolve(res.statusCode === 200 || res.statusCode === 401);
    });
    req.on('error', () => resolve(false));
    req.on('timeout', () => { req.destroy(); resolve(false); });
    req.end();
  });
}

function killProcessOnPort(port) {
  return new Promise((resolve) => {
    exec(`lsof -ti :${port}`, (err, stdout) => {
      if (err || !stdout.trim()) { resolve(); return; }
      const pids = stdout.trim().split('\n');
      let done = 0;
      for (const pid of pids) {
        exec(`kill -9 ${pid}`, () => {
          done++;
          if (done >= pids.length) resolve();
        });
      }
    });
  });
}

async function runTunnel(config) {
  let ssh = null;
  let consecutiveFailures = 0;
  let rebuildDelay = 2000;

  async function startSsh() {
    if (await httpHealthCheck(config.local_port)) {
      log(`Port ${config.local_port} is healthy, reusing existing tunnel.`);
      return;
    }

    log(`Port ${config.local_port} is not healthy, rebuilding tunnel...`);
    await killProcessOnPort(config.local_port);

    if (ssh) {
      try { ssh.kill('SIGKILL'); } catch {}
      ssh = null;
    }

    const args = buildSshArgs(config);
    log(`Starting SSH: ssh ${args.join(' ')}`);
    ssh = spawn('ssh', args, {
      stdio: ['ignore', 'ignore', 'pipe'],
      detached: false,
    });

    let stderrBuf = '';
    ssh.stderr.on('data', (d) => { stderrBuf += d; });

    await new Promise((r) => setTimeout(r, 1000));

    if (ssh.exitCode !== null || ssh.signalCode) {
      const hint = stderrBuf.includes('Address already in use')
        ? 'Local port is still occupied after cleanup. Try changing the Local Port in Settings (e.g., 18081).'
        : stderrBuf.includes('Connection refused')
        ? 'Remote SSH server refused the connection. Check SSH Host and Port.'
        : stderrBuf.includes('Permission denied') || stderrBuf.includes('authentication')
        ? 'SSH authentication failed. Check your SSH key or user name.'
        : '';
      log(`SSH exited immediately. ${hint}`);
      throw new Error(`SSH tunnel exited immediately. ${stderrBuf} ${hint}`);
    }

    log('SSH tunnel process started');
  }

  await startSsh();

  // Write PID file
  fs.writeFileSync(PID_PATH, String(process.pid));

  // Health check loop
  const interval = setInterval(async () => {
    if (!ssh) return;
    const ok = await httpHealthCheck(config.local_port);
    if (ok) {
      if (consecutiveFailures > 0) {
        log('Tunnel health check recovered');
      }
      consecutiveFailures = 0;
      rebuildDelay = 2000;
      return;
    }

    consecutiveFailures++;
    log(`Tunnel health check failed (${consecutiveFailures}/2)`);

    if (consecutiveFailures >= 2) {
      log('Tunnel deemed dead, auto-rebuilding...');
      consecutiveFailures = 0;
      await new Promise((r) => setTimeout(r, rebuildDelay));
      try {
        await startSsh();
        log('Tunnel rebuilt successfully');
        rebuildDelay = 2000;
      } catch (e) {
        log(`Tunnel rebuild failed: ${e.message}`);
        rebuildDelay = Math.min(rebuildDelay * 2, 60000);
      }
    }
  }, 10000);

  // Cleanup on exit
  function cleanup() {
    clearInterval(interval);
    if (ssh) {
      try { ssh.kill('SIGKILL'); } catch {}
    }
    try { fs.unlinkSync(PID_PATH); } catch {}
    process.exit(0);
  }
  process.on('SIGINT', cleanup);
  process.on('SIGTERM', cleanup);
  process.on('exit', cleanup);
}

function startDaemon(config) {
  const out = fs.openSync(LOG_PATH, 'a');
  const err = fs.openSync(LOG_PATH, 'a');
  const child = spawn(process.argv[0], [__filename, 'start'], {
    detached: true,
    stdio: ['ignore', out, err],
  });
  child.unref();
  fs.writeFileSync(PID_PATH, String(child.pid));
  console.log(`Tunnel daemon started (PID ${child.pid}). Logs: ${LOG_PATH}`);
  process.exit(0);
}

function stopTunnel() {
  try {
    const pid = fs.readFileSync(PID_PATH, 'utf8').trim();
    try {
      process.kill(parseInt(pid, 10), 'SIGTERM');
    } catch (e) {
      console.log('No tunnel process found.');
    }
    fs.unlinkSync(PID_PATH);
    console.log('Tunnel stopped.');
  } catch {
    console.log('No tunnel PID file found.');
  }
  process.exit(0);
}

function showStatus() {
  try {
    const pid = fs.readFileSync(PID_PATH, 'utf8').trim();
    try {
      process.kill(parseInt(pid, 10), 0);
      console.log(`Tunnel is running (PID ${pid})`);
    } catch {
      console.log('Tunnel PID file exists but process is dead.');
    }
  } catch {
    console.log('Tunnel is not running.');
  }
  process.exit(0);
}

async function main() {
  const cmd = process.argv[2];

  if (cmd === 'config') {
    await setupConfig();
    return;
  }

  if (cmd === 'status') {
    showStatus();
    return;
  }

  if (cmd === 'stop') {
    stopTunnel();
    return;
  }

  if (cmd === 'daemon') {
    const config = loadConfig();
    if (!config) {
      console.error('No config found. Run: node scripts/tunnel.js config');
      process.exit(1);
    }
    startDaemon(config);
    return;
  }

  if (cmd === 'start' || !cmd) {
    const config = loadConfig();
    if (!config) {
      console.error('No config found. Run: node scripts/tunnel.js config');
      process.exit(1);
    }
    log('CSM Tunnel Manager starting...');
    await runTunnel(config);
    return;
  }

  console.log(`Unknown command: ${cmd}`);
  console.log('Usage: node scripts/tunnel.js [start|daemon|stop|status|config]');
  process.exit(1);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
